### PR TITLE
add support for process classes to QueryBuilder.append

### LINKED ISSUE #2400: Add support for `Process` classes to `QueryBuilder.append`
If I have the `CalcJob` class `PwCalculation`, which has entry point `aiida.calculations:quantumespresso.pw`, calling `QueryBuilder.append(PwCalculation).all()` should tell the query builder to return all nodes that have a type string `node.process.calculation.calcjob.CalcJobNode.` and a `process_type` string `aiida.calculations:quantumespresso.pw`. This is essentially an `AND` statement of these two values. Note that since `append` supports adding tuples of classes, e.g. `QueryBuilder.append((PwCalculation, ArithmeticAddCalculation))`, this should be unwrapped into two sub filters that are `OR`'ed, i.e. in pseudo SQL
```
        WHERE (
            type = 'node.process.calculation.calcjob.CalcJobNode.' AND
            process_type = 'aiida.calculations:quantumespresso.pw'
        ) OR (
             type = 'node.process.calculation.calcjob.CalcJobNode.' AND
             process_type = 'aiida.calculations:arithmetic.add'
        )
```
technically since the `process_type` should also support querying for sub classes, the filter on process type should be an `OR` of `process_type = 'aiida.calculations:quantumespresso.pw'` and `process_type like 'aiida.calculations:quantumespresso.pw.'`

### LINKED ISSUE #2418: forbidden to pass list of classes to QueryBuilder
On `provenance_redesign`, I cannot pass a list of classes to the QueryBuilder:

```
In [1]: qb=QueryBuilder()

In [2]: qb.append([Node,Node])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-e54667174ba5> in <module>()
----> 1 qb.append([Node,Node])

/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/orm/querybuilder.py in append(self, cls, type, tag, filters, project, subclassing, edge_tag, edge_filters, edge_project, outerjoin, **kwargs)
    568
    569             else:
--> 570                 self._cls_to_tag_map[cls] = tag
    571                 l_class_added_to_map = True
    572

TypeError: unhashable type: 'list'
```

Using a tuple instead seems to work fine
```
In [1]: qb=QueryBuilder()

In [2]: qb.append((Node,Node))
Out[2]: <aiida.orm.querybuilder.QueryBuilder at 0x112105350>
```

Should lists be allowed?
If yes, this should be fixed and tested.
If not, we should change the documentation of `QueryBuilder.append`

### PR DESCRIPTION
**Pull request by ltalirz**
fix #2400 
fix #2418

 * move classifiers from individual return arguments into a `classifiers` dictionary
 * remove `query_type_string` from classifiers (can be deduced from `ormclass_type_string`)
 * add `process_type_string`
 * add basic tests for correct classification when passing a process class
 * allow passing lists (and sets) of classes to the Querybuilder (instead of just tuples)

To do
 - [x] write test: actually run a process (or create corresponding instances by hand) and query for it
 - [x] fix problem in building the docs
 - [x] document that one can pass process classes to the querybuilder (and what happens when you do this)

### CODE REVIEW

**Comment by ltalirz** [comment]
@sphuber This PR is now basically ready for review.

There are some strange test failure, only on python2.7
https://travis-ci.org/aiidateam/aiida_core/jobs/486559586#L3074

Just wondering whether you've come across something like this before I start investigating.

**Comment by sphuber** [comment]
@ltalirz I took the liberty to squash and rebase your commits onto `provenance_redesign` since I merged a whole slew of big PRs today. Just in case [I made a backup of it first on your remote](https://github.com/ltalirz/aiida_core/tree/issue_2400_process_in_querybuilder_backup). I also did this because I wanted to test it and on a first look it seems to work nicely.

I have two questions still:
1) I was a bit suprised to find that I could now add a tuple of two wildly different node types, for example `QueryBuilder().append((ArithmeticAddCalculation, ParameterData)).all()` will not complain and will retrieve all the nodes that match those classes. Do you know whether this was always the case?
2) When originally discussing this with @giovannipizzi , we though it would also be nice to allow sub classing on the process class, i.e. on the `process_type`. Imagine the following calculation entry points:
```
 'quantumespresso.pw': 'aiida_quantumespresso.workflows.pw:PwWorkChain',
 'quantumespresso.pw.base': 'aiida_quantumespresso.workflows.pw.base:PwBaseWorkChain',
 'quantumespresso.pw.relax': 'aiida_quantumespresso.workflows.pw.relax:PwRelaxWorkChain',
```
The entry point names here could function just as the node type strings, as in when appending the `PwWorkChain` class, it would not only match `process_type=quantumespresso.pw` but also `process_type LIKE 'quantumespresso.pw.%'`, which would therefore also find the instances of `PwBaseWorkChain` and `PwRelaxWorkChain`. If this is complicated, it would be fine to add this later in `v1.0.0`, but I mention it here if it might require a fundamental change in the current solution that we might want to consider before merging

**Comment by ltalirz** [comment]
>  1. Do you know whether this was always the case?

I think it works as long as what you are querying for resolves to "Nodes", which is the case in your example. I didn't modify the logic for handling lists - if you would like to modify it, let me know.
If it was up to me, I would actually vote to remove this functionality completely, since I got the impression that 40% of the querybuilder code just deals with distinguishing between what to do if someone is passing a list or not. But maybe there are important use cases that I am not aware of.

>  2. It would also be nice to allow sub classing on the process class

That's a valid point. I skipped this for the moment, since I didn't know whether there was any kind of standardization of the `process_type` string (and whether these strings were still going to change). 
Note that at the moment it can be e.g. `aiida.work.workchain.WorkChain` (no colon) but also `aiida.calculations:arithmetic.add` (with colon).
If you give me a specification of the `process_type` string and a foolproof rule for the subclassing query, it should be very easy for me to add.

The only thing is that there is currently only one `subclassing` parameter but I guess we could live with this applying to both the ormclass type and the process type simultaneously.

**Comment by ltalirz** [comment]
So, to sum up:

 1. the process type string is either
    * the import path of the class that defines the process, or
    * the entry point of the class defining the process (if there is one). Uses `get_entry_point_from_class`
 2. the type string is either
    * the import path of the defining class, appended by '.'
    * the entry point of the class defining the class (if there is one), appended by '.'.
      Uses `get_entry_point_string_from_class`, which adds a colon (different from `get_entry_point_from_class`)

 3. `subclassing` will enable subclassing both on the ormclass type and the process type
    * for the process type, we will only enable subclassing, when there is an entry point (this was suggested by @giovannipizzi . Not sure whether this is a good idea, since e.g. we might want subclassing also for AiiDA-internal process types, let me know)

@sphuber Is this correct, and is that how we should proceed?

Looking at this mess, I think users will rightfully ask why it has to be so complicated. I'm ok to go with this for the moment, though, to get the 1.0 beta out.

**Comment by sphuber** [comment]
For 2. the type string is never an entry point string but always one that is based on the module path of the node class. For `ProcessNodes` and sub classes this is easy because they are always internal to `aiida-core` and cannot be sub classed. `Data` sub classes can be defined externally and their type string is based on their entry point name if they have an associated entry point, with a fallback on the full module path plus class name.

For the `process_type`, we only will support querying if the `Process` class passed to `append` has a matching entry point, and if not an exception should be raised explaining that querying for unregistered process classes is not supported. This is with the exception of the internal classes `Process`, `CalcJob` and `WorkChain` where there will only be a filter on the `type` string and no filter at all on the `process_type` string.

3. Yes, `subclassing` means all or nothing

**Comment by ltalirz** [comment]
> This is with the exception of the internal classes `Process`, `CalcJob` and `WorkChain` where there will only be a filter on the `type` string and no filter at all on the `process_type` string.

Hm... in the end I don't think it is a good rule. What if someone specifies only a process_type_string (and no orm class type)? Instead of using no filter, for `Process`, `CalcJob` and `WorkChain` I will simply enable the same subclassing as one does for the orm class type string.

**Comment by sphuber** [comment]
You mean you will enable the same subclassing on the `process_type` as you would on the `type` string when the user passes `Process`, `CalcJob` or `WorkChain`? That does not make sense, because their process type is empty and if it would be done on a module path based string, subclasses would have to be put below the base class in the module hierarchy for it to work, which means users have to place their subclasses in the `aiida-core` tree, which is not something we want. Or am I missing something here?

**Comment by ltalirz** [comment]
> when the user passes `Process`, `CalcJob` or `WorkChain`? That does not make sense, because their process type is empty

The process_type of `Process` is: 
```
In [4]: Process.get_process_type()
Out[4]: 'aiida.work.processes.Process'
```
This is what would be passed to the QueryBuilder

> if it would be done on a module path based string, subclasses would have to be put below the base class in the module hierarchy for it to work, which means users have to place their subclasses in the `aiida-core` tree, which is not something we want.

I was thinking more of the usecase where you want to catch aiida-internal subclasses.
But you are right - one might want to append a `WorkChain` and get all possible workchains from all plugins.
This is more important than the more esoteric case of a process_type without an ormclass type.

**Comment by ltalirz** [comment]
Hm... after some digging, it turns out that `reset_manager()` is actually called not only in tearDownClass but also in tearDown, i.e. between every test.
I suspect that *somewhere* there is some cache that is not cleared, which gives rise to these problems.

For the record: Deleting the travis cache did change something (perhaps it changes some order):
 * before: envs failing were django 2.7, sqla 2.7/3.6; tests: test_persisting, test_report_dbloghandler
 * after: envs failing are django 2.7/3.6, sqla 2.7; tests: test_persisting, test_str (2x), test_report_dbloghandler (1x)

**Comment by sphuber** [comment]
Are you sure you are on the latest `provenance_design` commit. I fixed a bug related to the `DbLogHandler` recently that could render the database in an inconsistent state, at least for SqlA. This is definitely the error I see in `test_report_dbloghandler` on SqlA 3.6 that fails

**Comment by ltalirz** [comment]
Just an update - I'm finally able to reproduce the tests locally (not when running `verdi devel tests db.work` but when running all tests with `verdi devel tests`).
Will try to pin down what causes the problem.

**Comment by ltalirz** [comment]
Indeed `db.query` seems to be it (I had added new tests there).
To reproduce: `verdi devel tests db.query db.work.work_chain -v`

**Comment by ltalirz** [comment]
Ok, the culprit seems to be this test that is running a workflow (when I comment it out, the tests pass):
https://github.com/aiidateam/aiida_core/pull/2421/files#diff-0f40137d00b106fee080eb5ce04a0ff6R114

**Comment by ltalirz** [comment]
This is not yet the solution, but it may be worth noticing that the test class of this test does `self.clean_db()` in the `setUp` which is unusual (e.g. this is not the case in the test classes used for the workchain tests).

I'll stop working on this for a few hours, comments welcome

**Comment by ltalirz** [comment]
@sphuber We discussed at some point you wanted to use `ProcessNode._set_process_type` to se the process type. It turns out, however, that there is a `ProcessNode.process_type` with a getter and a setter.

I therefore propose to get rid of `ProcessNode._set_process_type`, do you agree?
Either this or the `process_type` property should be moved out of `Node` and into `ProcessNode`.

**Comment by sphuber** [comment]
If it is a proper setter, as in it gets a value and sets it, then it is fine to get rid of  `_set_process_type`. If I remember correctly, you moved the logic that was in there in a `build_process_type`. The engine can then call `node.process_type = ProcessClass.build_process_type()` or something like that.

**Comment by ltalirz** [comment]
@sphuber @giovannipizzi I believe I finally found the cause of the failing tests.

The problem is that WorkChain **classes** can store references to ORM objects, such as [here](https://github.com/aiidateam/aiida_core/blob/b22f55155e7a602ac42b00edf62e57e4a458f55b/aiida/backends/tests/work/test_work_chain.py#L94).
After a database reset these classes will have references to orphaned nodes, so in general one cannot touch any WorkChain that was imported before.

While it is easy to avoid this in this PR, I think it is a serious issue.
For example, it means WorkChains will prevent us from switching backends in AiiDA scripts because the Data nodes they already reference won't adapt to the backend we set.
Has this been considered so far?

**Comment by ltalirz** [comment]
Tests are passing now; also added some (relatively minimal) documentation.

**Comment by giovannipizzi** [comment]
actually, one is still failing...

**Comment by ltalirz** [comment]
One of these "no output for 10 minutes" cases.
Actually, the same test (sqla with python 3.6) was also failing for the same reason [on the previous commit](https://travis-ci.org/aiidateam/aiida_core/builds/496669089?utm_source=github_status&utm_medium=notification). After resetting the cache and restarting the job, it passes. Not quite sure what to make of this.

Also, should I open an issue for [the problem mentioned above](https://github.com/aiidateam/aiida_core/pull/2421#issuecomment-466084696)?

**Comment by sphuber** [comment]
I think the problem of node references is something that has caused problems before, for example it was also related to the creation of duplicate node UUIDs. I think we should maybe disallow adding Node instances as defaults in the classmethods like the `define` and we should replace this with a lambda that will not instantiate the node until it is actually called, not directly on import. Anyway, this is a bigger problem and deserves its own issue and PR, I would not bother with it here.

**Comment by ltalirz** [comment]
Ok, I'll open an issue for it.

**Comment by ltalirz** [comment]
closing until final merge to avoid blocking builds

File: aiida/engine/processes/process.py [hunk 1]
```diff
@@ -704,3 +729,17 @@ def _get_namespace_list(namespace=None, agglomerate=True):
             split_ns = namespace.split('.')
             namespace_list.extend(['.'.join(split_ns[:i]) for i in range(1, len(split_ns) + 1)])
         return namespace_list
+
+
+def get_query_string_from_process_type_string(process_type_string):  # pylint: disable=invalid-name
+    """
+    Take the type string of a Node and create the queryable type string
```
Comment (sphuber) [inline aiida/engine/processes/process.py]: Take the _process_ type string

File: aiida/cmdline/commands/cmd_daemon.py [hunk 2]
```diff
@@ -211,7 +211,7 @@ def restart(ctx, reset, no_wait):
 @click.option('--foreground', is_flag=True, help='Run in foreground.')
 @decorators.with_dbenv()
 @decorators.check_circus_zmq_version
-def start_circus(foreground):
+def start__pcircus(foreground):
```
Comment (sphuber) [inline aiida/cmdline/commands/cmd_daemon.py]: why the name change? Does this accomplish the hiding of the command? I found that `click==7.0.0` has implemented a `hidden` keyword for `commands` which works, I just couldn't get the bash completion to work though. It is also supposed to hide it from autocompletion. In any case, I would leave this untouched and do a proper click based solution when we get it to work

File: aiida/orm/nodes/process/process.py [hunk 3]
```diff
@@ -104,17 +103,7 @@ def set_process_type(self, process_class):
...
-
-        # If the process is a registered plugin the corresponding entry point will be used as process type
-        process_type = get_entry_point_string_from_class(class_module, class_name)
-
-        # If no entry point was found, default to fully qualified path name
-        if process_type is None:
-            self.process_type = '{}.{}'.format(class_module, class_name)
-        else:
-            self.process_type = process_type
+        self.process_type = process_class.build_process_type()
```
Comment (sphuber) [inline aiida/orm/nodes/process/process.py]: I would instead delegate the building of the process type itself to the `Process` instance that actually calls `set_process_type`. The setter here then simply takes the value it receives as is and assigns it (with optional validation but don't think we had any or need any now)

File: aiida/plugins/entry_point.py [hunk 4]
```diff
@@ -275,6 +275,7 @@ def get_entry_point_string_from_class(class_module, class_name):
     by splitting on the separator, which will give the group and entry point, which should
     the corresponding factory to uniquely determine and load the class
 
+    :rtype:
```
Comment (sphuber) [inline aiida/plugins/entry_point.py]: incomplete

File: aiida/orm/querybuilder.py [hunk 5]
```diff
@@ -53,117 +53,211 @@
...
+    import warnings
+
+    value = classifiers['process_type_string']
+
+    if not subclassing:
+        filter = {'==': value}
+    else:
+        if ":" in value:
+            # if value is an entry point, do usual subclassing
+            filter = {'or': [
```
Comment (sphuber) [inline aiida/orm/querybuilder.py]: Why do we need an `or` statement again in the case of sub classing on the `process_type` but not in the case of the `type`? I forgot

File: aiida/cmdline/commands/cmd_daemon.py (hunk 2 above)
Comment (ltalirz) [reply aiida/cmdline/commands/cmd_daemon.py]: This somehow popped out of the merge - it's not like I wanted to add this.
Will revert the name

File: aiida/orm/querybuilder.py (hunk 5 above)
Comment (ltalirz) [reply aiida/orm/querybuilder.py]: The `process_type` is stored in the database without a `.` at the end.

Comment (sphuber) [reply aiida/orm/querybuilder.py]: Ah right, if you push the small changes for the other comments I will approve and merge

Comment (ltalirz) [reply aiida/orm/querybuilder.py]: I've addressed your comments now.
Your comment made me realize an issue - it seems to me one should use `get_query_type_from_type_string` also [here](https://github.com/aiidateam/aiida_core/pull/2421/files#diff-ee49942ec9ece0e5ff43f3572bb42843R205) but when adding it, some tests fail.
I currently don't understand why (perhaps I'm also mistaken). If it is an issue, I'll prefer to fix this in a separate PR.

Comment (sphuber) [reply aiida/orm/querybuilder.py]: Yes, let's get this merged and address any issues that may arise later

**Comment by sphuber** [comment]
Thanks a lot @ltalirz , great work!

**Pull request merged**